flink-libraries/flink-table/src/main/java/org/apache/calcite/sql/type/SqlTypeFactoryImpl.java

```diff
@@ -0,0 +1,575 @@
+		assert types.size() >= 1;
+
+		RelDataType type0 = types.get(0);
+		if (type0.getSqlTypeName() != null) {
+			RelDataType resultType = leastRestrictiveSqlType(types);
+			if (resultType != null) {
+				return resultType;
+			}
+			return leastRestrictiveByCast(types);
+		}
+
+		return super.leastRestrictive(types);
+	}
+
+	private RelDataType leastRestrictiveByCast(List<RelDataType> types) {
+		RelDataType resultType = types.get(0);
+		boolean anyNullable = resultType.isNullable();
+		for (int i = 1; i < types.size(); i++) {
+			RelDataType type = types.get(i);
+			if (type.getSqlTypeName() == SqlTypeName.NULL) {
+				anyNullable = true;
+				continue;
+			}
+
+			if (type.isNullable()) {
+				anyNullable = true;
+			}
+
+			if (SqlTypeUtil.canCastFrom(type, resultType, false)) {
+				resultType = type;
+			} else {
+				if (!SqlTypeUtil.canCastFrom(resultType, type, false)) {
+					return null;
+				}
+			}
+		}
+		if (anyNullable) {
+			return createTypeWithNullability(resultType, true);
+		} else {
+			return resultType;
+		}
+	}
+
+	// implement RelDataTypeFactory
+	public RelDataType createTypeWithNullability(
+		final RelDataType type,
+		final boolean nullable) {
+		RelDataType newType;
+		if (type instanceof BasicSqlType) {
+			BasicSqlType sqlType = (BasicSqlType) type;
+			newType = sqlType.createWithNullability(nullable);
+		} else if (type instanceof MapSqlType) {
+			newType = copyMapType(type, nullable);
+		} else if (type instanceof ArraySqlType) {
+			newType = copyArrayType(type, nullable);
+		} else if (type instanceof MultisetSqlType) {
+			newType = copyMultisetType(type, nullable);
+		} else if (type instanceof IntervalSqlType) {
+			newType = copyIntervalType(type, nullable);
+		} else if (type instanceof ObjectSqlType) {
+			newType = copyObjectType(type, nullable);
+		} else {
+			return super.createTypeWithNullability(type, nullable);
+		}
+		return canonize(newType);
+	}
+
+	private void assertBasic(SqlTypeName typeName) {
+		assert typeName != null;
+		assert typeName != SqlTypeName.MULTISET
+			: "use createMultisetType() instead";
+		assert !SqlTypeName.INTERVAL_TYPES.contains(typeName)
+			: "use createSqlIntervalType() instead";
+	}
+
+	private RelDataType leastRestrictiveSqlType(List<RelDataType> types) {
+		RelDataType resultType = null;
+		int nullCount = 0;
+		int nullableCount = 0;
+		int javaCount = 0;
+		int anyCount = 0;
+
+		for (RelDataType type : types) {
+			final SqlTypeName typeName = type.getSqlTypeName();
+			if (typeName == null) {
+				return null;
+			}
+			if (typeName == SqlTypeName.ANY) {
+				anyCount++;
+			}
+			if (type.isNullable()) {
+				++nullableCount;
+			}
+			if (typeName == SqlTypeName.NULL) {
+				++nullCount;
+			}
+			if (isJavaType(type)) {
+				++javaCount;
+			}
+		}
+
+		//  if any of the inputs are ANY, the output is ANY
+		if (anyCount > 0) {
+			// for flink, it should return GenericRelDataType
+			if (types.get(0) instanceof GenericRelDataType &&
```

Comment:
I modified here to support FLink' ANY type

Reply:
We shouldn't copy Calcite code and add Flink dependencies to it. This class can never be removed in a later Calcite release. Isn't there a cleaner way to fix this issue. Please also add a header to this file (as we did it with other Calcite files), to add information when we can remove it again from the Flink code base.